commit babef57ae733136084b7c72d28c365a038cb4f46 (refs/changes/54/1454/8)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2017-07-26T21:47:14+00:00 (2 years, 2 months ago)
    
    MORAY-389 Limit which modules triggers are allowed to require
    Reviewed by: David Pacheco <dap@joyent.com>
    Reviewed by: Patrick Mooney <patrick.mooney@joyent.com>

diff --git a/lib/errors.js b/lib/errors.js
index 757e68a..a2c94b1 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -183,6 +183,18 @@ function InvalidIndexTypeError(cause, name, type) {
 util.inherits(InvalidIndexTypeError, WError);
 
 
+function InvalidRequireError(cause, name) {
+    if (arguments.length === 1) {
+        name = cause;
+        cause = {};
+    }
+    assert.string(name, 'name');
+    WError.call(this, cause, '"%s" is not a permitted module', name);
+    this.name = this.constructor.name;
+}
+util.inherits(InvalidRequireError, WError);
+
+
 function InvalidQueryError(cause, filter) {
     if (arguments.length === 1) {
         filter = cause;
@@ -331,6 +343,7 @@ module.exports = {
     InvalidBucketNameError: InvalidBucketNameError,
     InvalidIndexDefinitionError: InvalidIndexDefinitionError,
     InvalidIndexTypeError: InvalidIndexTypeError,
+    InvalidRequireError: InvalidRequireError,
     InvalidQueryError: InvalidQueryError,
     InvocationError: InvocationError,
     NoDatabaseError: NoDatabaseError,
diff --git a/lib/objects/common.js b/lib/objects/common.js
index 85bab9c..1f6042a 100644
--- a/lib/objects/common.js
+++ b/lib/objects/common.js
@@ -18,6 +18,7 @@ var LRU = require('lru-cache');
 var once = require('once');
 var vasync = require('vasync');
 var clone = require('clone');
+var vm = require('vm');
 
 var dtrace = require('../dtrace');
 
@@ -25,6 +26,7 @@ var mod_errors = require('../errors');
 var BucketNotFoundError = mod_errors.BucketNotFoundError;
 var NotIndexedError = mod_errors.NotIndexedError;
 var InvalidIndexTypeError = mod_errors.InvalidIndexTypeError;
+var InvalidRequireError = mod_errors.InvalidRequireError;
 var InvalidQueryError = mod_errors.InvalidQueryError;
 var EtagConflictError = mod_errors.EtagConflictError;
 
@@ -691,6 +693,55 @@ function stdOutput(req) {
 }
 
 
+/*
+ * Some Moray triggers used the same "microtime" and "crc" modules that
+ * Moray used. We allow those triggers to continue working by passing
+ * our own require() function into their environment and restricting them
+ * to only loading those modules.
+ */
+function guardedRequire(name) {
+    if (name !== 'microtime' && name !== 'crc') {
+        throw new InvalidRequireError(name);
+    }
+
+    return require(name);
+}
+
+
+function evalTrigger(f) {
+    /*
+     * We evaluate triggers in their own environment, to avoid leaking
+     * functions and variables from our own. Several things are passed
+     * into the environment of the evaluated triggers:
+     *
+     * - Error, so that all returned errors will be 'instanceof Error'
+     * - require, to allow backwards-compatibility for triggers that
+     *   used Moray's modules.
+     */
+    var ctx = {
+        Error: Error,
+        require: guardedRequire
+    };
+
+    vm.runInNewContext('fn = ' + f, ctx, 'moray-trigger');
+
+    assert.func(ctx.fn, 'trigger');
+
+    /*
+     * We store the trigger in an object so that when it gets called
+     * its 'this' value points to our object instead of to the scope
+     * in which it was called.
+     */
+    var trigger = { run: ctx.fn };
+
+    function safetyNet(arg, cb) {
+        trigger.run(arg, once(cb));
+    }
+
+    return (safetyNet);
+}
+
+
 function loadBucket(req, cb) {
     assert.object(req, 'req');
     assert.object(req.bucket, 'req.bucket');
@@ -743,21 +794,12 @@ function loadBucket(req, cb) {
         if (!row) {
             cb(new BucketNotFoundError(req.bucket.name));
         } else {
-            function parseFunctor(f) {
-                var fn;
-                assert.ok(eval('fn = ' + f));
-                function safetyNet(arg, _cb) {
-                    return (fn(arg, once(_cb)));
-                }
-                return (safetyNet);
-            }
-
             var r = row;
             req.bucket = {
                 name: r.name,
                 index: JSON.parse(r.index),
-                pre: JSON.parse(r.pre).map(parseFunctor),
-                post: JSON.parse(r.post).map(parseFunctor),
+                pre: JSON.parse(r.pre).map(evalTrigger),
+                post: JSON.parse(r.post).map(evalTrigger),
                 options: JSON.parse(r.options || {}),
                 mtime: new Date(r.mtime)
             };
@@ -1061,6 +1103,7 @@ module.exports = {
     buildWhereClause: buildWhereClause,
     cacheKey: cacheKey,
     checkEtag: checkEtag,
+    evalTrigger: evalTrigger,
     loadBucket: loadBucket,
     shootdownBucket: shootdownBucket,
     verifyBucket: verifyBucket,
